feat(python): expose MERGE operation

ion-elgreco opened the pull request:
# Description
This exposes MERGE commands to the Python API.  The updates and predicates are first kept in the Class TableMerger and only dispatched to Rust after `TableMerge.execute()`. 

This was my first thought on how to implement it since I have limited experience with Rust and PyO3 (still learning 😄). Maybe a more elegant solution is that every class method on TableMerger is dispatched to Rust and then the Rust MergeBuilder gets serialized and sent back to Python (back and forth). Let me know your thoughts on this. If this is better, I could also do this in the next PR, so we at least can push this one out sooner.

Couple of issues at the  moment, I need feedback on, where the first one is blocking since I can't test it now:

~- Source_alias is not applying, somehow during a schema check the prefix is missing, however when I printed the lines inside merge, it showed the prefix correctly. So not sure where the issue is~
~-  I had to make datafusion_utils public since I needed to get the Expression Struct from it, is this the right way to do that?  @Blajda~

Edit:
I will pull @Blajda's changes https://github.com/delta-io/delta-rs/pull/1705 once merged with develop: 


# Related Issue(s)
<!---
For example:

- closes #106
--->
closes  #1357

Blajda commented:
@ion-elgreco Overall looks good. Please add some python tests for each of the various operations and assert data is updated as expected. When at least one test is added I'll take a look at the alias issue more in depth.

> I had to make datafusion_utils public since I needed to get the Expression Struct from it, is this the right way to do that?

Since only the utilities around `Expression` need to made public I think it would be better to promote them to `delta_datafusion.rs`. Eventually we need to refactor the file into smaller units.

ion-elgreco commented:
> @ion-elgreco Overall looks good. Please add some python tests for each of the various operations and assert data is updated as expected. When at least one test is added I'll take a look at the alias issue more in depth.
> 
> > I had to make datafusion_utils public since I needed to get the Expression Struct from it, is this the right way to do that?
> 
> Since only the utilities around `Expression` need to made public I think it would be better to promote them to `delta_datafusion.rs`. Eventually we need to refactor the file into smaller units.
> 
> 

@Blajda thanks! Will add the tests and ping you later

ion-elgreco commented:
@Blajda I added one test quickly where you can see the source_alias will not apply, however running it from terminal it gives also this error:` FAILED tests/test_merge.py::test_merge_when_matched_delete_wo_predicate - AttributeError: 'builtins.RawDeltaTable' object has no attribute 'merge_execute'`. I don't really see why this happens. In a Jupyter notebook I don't get this error, there it executes and will throw this error:

`DeltaError: Generic DeltaTable error: Schema error: No field named source.id. Valid fields are .id, .price, .sold, .deleted.`

_I will add the other tests later tonight after work._

Blajda commented:
@ion-elgreco I understand the root cause of the schema error. The issue is outlined in #1699 
Give me till the end of the week to resolve this.

ion-elgreco commented:
> @ion-elgreco I understand the root cause of the schema error. The issue is outlined in #1699 
> Give me till the end of the week to resolve this.

Thanks for looking into it! @Blajda 

Let me know once your fix is ready, then I'll check on my side if everything runs now :)

Blajda commented:
@ion-elgreco 
This PR contain the fix: #1705
I have one last TODO before removing the draft status but please try making a branch of your changes and merging the fix to see if everything works.

ion-elgreco commented:
> @ion-elgreco The source and target aliases should be optional arguments. If the column names of both relations are unique then specifying an alias is not required. The rust merge code has a test to demonstrate this.

Ah ok thanks, I will revert the change then!

wjones127 commented on python/src/lib.rs:
```diff
@@ -396,6 +397,185 @@ impl RawDeltaTable {
+    ))]
+    pub fn merge_execute(
+        &mut self,
+        source: PyArrowType<RecordBatch>,
```
I would instead take a `PyArrowType<Box<dyn RecordBatchReader + Send>>` here

wjones127 commented on python/src/lib.rs:
```diff
@@ -396,6 +397,185 @@ impl RawDeltaTable {
+        not_matched_by_source_delete_all: Option<bool>,
+    ) -> PyResult<String> {
+        let ctx = SessionContext::new();
+        let source_df = ctx.read_batch(source.0).unwrap();
```
Instead of `read_batch`, create a [MemTable](https://docs.rs/datafusion/latest/datafusion/datasource/memory/struct.MemTable.html) from the batches in the reader.

wjones127 reviewed (changes requested):
The main blocking change is the bug related to `pa.Table`. Make sure you are handling tables that contains multiple record batches. I've provided suggestions for how to handle that.

ion-elgreco commented on python/deltalake/table.py:
```diff
@@ -747,6 +781,309 @@ def delete(self, predicate: Optional[str] = None) -> Dict[str, Any]:
+        self.target_alias = target_alias
+        self.safe_cast = safe_cast
+        self.writer_properties: Optional[Dict[str, Optional[int]]] = None
+        self.matched_update_updates: Optional[Dict[str, str]] = None
+        self.matched_update_predicate: Optional[str] = None
```
I went off the examples in delta-spark, but we can apparently do multiple clauses in the Rust API. I think instead of passing it a value, can pass perhaps a list of dicts, having this structure:` { predicate: "x > 1", updates: {"col" : "col"}}`
And then add an iterator for each of those in rust. 

Should we get it in this PR or maybe I can add it in a smaller PR updating the logic afterwards?

wjones127 replied:
I'm fine making that a follow up, but thought I would point that out.

ion-elgreco replied:
Yes good for pointing it out! I will do that in a follow-up then.

ion-elgreco commented on python/tests/test_merge.py:
```diff
@@ -0,0 +1,460 @@
+    source_table = pa.table(
+        {
+            "id": pa.array(["5"]),
+            "price": pa.array([1], pa.int64()),
+            "sold": pa.array([1], pa.int32()),
+            "deleted": pa.array([False]),
```
Nope, good point, I will switch it up and remove or rename some columns here and there.

ion-elgreco commented on python/deltalake/table.py:
```diff
@@ -525,6 +525,40 @@ def optimize(
+        predicate in SQL query like format. You can also specify on what to do when the underlying data types do not
+        match the underlying table.
+
+        Args:
+            source (pyarrow.Table | pyarrow.RecordBatch): source data
+            predicate (str): SQL like predicate on how to merge
+            source_alias (str): Alias for the source table
+            target_alias (str): Alias for the target table
+            error_on_type_mismatch (bool): specify if merge will return error if data types are mismatching :default = True
+
+        Returns:
+            TableMerger: TableMerger Object
```
My bad, I am using autodocstring in VS code which was set to use Google Doc strings.

ion-elgreco commented on python/deltalake/table.py:
```diff
@@ -525,6 +525,40 @@ def optimize(
+        Returns:
+            TableMerger: TableMerger Object
+        """
+        if isinstance(source, pyarrow.Table):
+            source = source.to_batches()[0]
```
Thanks for pointing that out, I assumed it was always one batch because I was exporting some polars data frames to arrow.

ion-elgreco commented on python/tests/test_merge.py:
```diff
@@ -0,0 +1,460 @@
+    assert result == expected
+
+
+# # ## Add when_not_matched_by_source_delete_wo_predicate ?
```
Done!

wjones127 commented on python/deltalake/table.py:
```diff
@@ -525,6 +526,62 @@ def optimize(
+        elif isinstance(source, pyarrow.RecordBatch):
+            batch_iter = [source]
+        elif isinstance(source, pyarrow.Table):
+            batch_iter = source.to_batches()
```
I think for older PyArrow versions to work, this should be `to_reader()`:
```suggestion
            batch_iter = source.to_reader()
```

ion-elgreco replied:
As long as our minimum Pyarrowversion is 8, to_reader() is also fine. To_batches however seems to be since the very early version of Pyarrow. Not sure if there is a performance difference though.

wjones127 replied:
Yeah.

wjones127 commented on python/deltalake/table.py:
```diff
@@ -525,6 +526,62 @@ def optimize(
+            batch_iter = [source]
+        elif isinstance(source, pyarrow.Table):
+            batch_iter = source.to_batches()
+        else:
+            batch_iter = source
```
Should we instead error here? At this point it's not one of the three supported types?

ion-elgreco replied:
We should indeed,

wjones127 commented:
Great work on this!

ion-elgreco commented:
> Great work on this!

Thanks @wjones127!

XiphoidGrizzly commented:
Is there an ETA on this enhancement going GA?

ion-elgreco commented:
> Is there an ETA on this enhancement going GA?

It's available in the latest release v0.12.0

yefetBenTili commented:
@ion-elgreco Thanks for great work! should [docs ](https://delta-io.github.io/delta-rs/python/api_reference.html )be updated accordingly?

ion-elgreco commented:
@yefetBenTili Thanks! 

With regards to the docs, the usage page still doesn't mention the new merge functionality yet, are you willing to perhaps add that?

mohitbansal-gep commented:
Hi I was referring to this blog post for example, https://delta.io/blog/2023-10-22-delta-rs-python-v0.12.0/ 
however i found an issue that if the schema has columns with Camel Case then the merge operation throws Generic Delta table error

ion-elgreco commented:
@mohitbansal-gep if you're running into an issue please create an issue on the board:)
